src/btree/bt_sync.c

```diff
@@ -384,11 +384,8 @@ __sync_page_skip(
     if (!__wt_ref_addr_copy(session, ref, &addr))
         return (0);
 
-    /*
-     * Skip reading the pages that are normal leaf pages or don't have an aggregated durable stop
-     * timestamp.
-     */
-    if (addr.type == WT_ADDR_LEAF_NO || addr.ta.newest_stop_durable_ts == WT_TS_NONE) {
+    /* Skip reading normal leaf pages. */
```

Comment:
This means that checkpoint cleanup will always try to read all internal pages in a tree.

Is there some other mechanism we can use to determine if there are any deleted children of an internal page? It would be nice if cleanup didn't *have* to read parts of the tree it doesn't need to.

Reply:
Yes, that's correct it reads all the internal pages. 
1. We have an aggregated timestamp to decide whether to read or not.
2. The Main problem is with the transaction id. In a running database, checking the transaction id is not a problem, it works.  Upon restart, there is no way to identify whether an internal page is having the deleted pages to read it, as both timestamp and transaction id are zero.

Reply:
There is aggregated data. Following aggregated members track the stop information. 
`newest_stop_durable_ts` - Contains the aggregated remove durable timestamp.
`newest_stop_ts` - Contains the aggregated remove commit timestamp including WT_TS_MAX if there are no removed operations.
`newest_stop_txn` - Contains the aggregated remove commit transaction id including WT_TXN_MAX if there are no removed operations.

As part of the checkpoint cleanup, we want to decide whether an internal page can be skipped or read back into the cache to check for obsolete pages.

The aggregated data of `newest_stop_ts` and `newest_stop_txn` can contain WT_TS_MAX and WT_TXN_MAX as values, so relying on them to read/skip the page can lead to either reading all pages or don't read anything.

So the existing logic uses to check for `newest_stop_durable_ts` as having any value or not. If there is a stop, there could be possibly an obsolete page, so do not skip the page. If it is a non-timestamped table, this logic is causing problems as it skips the page to read into the cache.

we do not have any `newest_stop_txn` without including WT_TXN_MAX, like durable timestamps. Even if we have, we reset the transaction ids after the server restart including the aggregated transaction ids. We do not have `write generation number` as part of the address cell in deciding whether to consider the transaction ids or not. 

Due to the above problem, I removed the check for the durable timestamp and read all the internal pages. 

For the longer term, I feel we need to change the aggregated time window information that we track. The `oldest_start_ts` we track is not used anywhere in the code. Maybe changing it to track the required information would be good.

Reply:
How about the following approach of allowing reading the internal pages?
1. Read all the internal pages for logged tables. Logged tables do not support timestamps.
2. Read only the pages that have the aggregated stop timestamp for non-logged tables.

The problem with the above approach is that if the truncate is performed without a timestamp on the non-logged tables, still they can be missed.

Reply:
Your suggestion sounds good to me @kommiharibabu . I think we should also use this as motivation for getting WT-7788 (improving checkpoint cleanup efficiency) back on the radar.

Reply:
Thanks. Fixed the code as per the above. Yes, we need to prioritize WT-7788 to improve the checkpoint cleanup efficiency.